internal/batch/batch_processor.go

```diff
@@ -101,103 +130,254 @@ func (bw *batchWork) estimateSize() int64 {
+	}
+}
+
+func (bp *batchProcessor) newAssembly(initalWork ...*batchWork) {
+	bp.assemblyID = fftypes.NewUUID()
+	bp.assemblyQueue = append([]*batchWork{}, initalWork...)
+	bp.assemblyQueueBytes = batchSizeEstimateBase
+}
+
+// addWork adds the work to the assemblyQueue, and calculates if we have overflowed with this work.
+// We check for duplicates, and add the work in sequence order.
+// This helps in the case for parallel REST APIs all committing to the DB at a similar time.
+// With a sufficient batch size and batch timeout, the batch will still dispatch the messages
+// in DB sequence order (although this is not guaranteed).
+func (bp *batchProcessor) addWork(newWork *batchWork) (full, overflow bool) {
+	newQueue := make([]*batchWork, 0, len(bp.assemblyQueue)+1)
+	added := false
+	// Check it's not in the recently flushed list
+	for _, flushedSequence := range bp.flushedSequences {
+		if newWork.msg.Sequence == flushedSequence {
+			log.L(bp.ctx).Debugf("Ignoring add of recently flushed message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			return false, false
+		}
+	}
+	// Build the new sorted work list, checking there for duplicates too
+	for _, work := range bp.assemblyQueue {
+		if newWork.msg.Sequence == work.msg.Sequence {
+			log.L(bp.ctx).Debugf("Ignoring duplicate add of message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			return false, false
+		}
+		if !added && newWork.msg.Sequence < work.msg.Sequence {
+			newQueue = append(newQueue, newWork)
+			added = true
+		}
+		newQueue = append(newQueue, work)
+	}
+	if !added {
+		newQueue = append(newQueue, newWork)
+	}
+	log.L(bp.ctx).Debugf("Added message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+	bp.assemblyQueueBytes += newWork.estimateSize()
+	bp.assemblyQueue = newQueue
+	full = len(bp.assemblyQueue) >= int(bp.conf.BatchMaxSize) || (bp.assemblyQueueBytes >= bp.conf.BatchMaxBytes)
+	overflow = len(bp.assemblyQueue) > 1 && (bp.assemblyQueueBytes > bp.conf.BatchMaxBytes)
+	return full, overflow
+}
+
+func (bp *batchProcessor) addFlushedSequences(flushAssembly []*batchWork) {
+	// We need to keep track of the sequences we're flushing, because until we finish our flush
+	// the batch processor might be re-queuing the same messages to use due to rewinds.
+
+	// We keep twice the batch size, which might be made up of multiple batches
+	maxFlushedSeqLen := int(2 * bp.conf.BatchMaxSize)
+
+	// We keep as much of the END of the existing set as we can, and shift it forwards.
+	// Then we add the whole of the current flush set after that
+	combinedLen := len(flushAssembly) + len(bp.flushedSequences)
+	newLength := combinedLen
+	if combinedLen > maxFlushedSeqLen {
+		newLength = maxFlushedSeqLen
+	}
+	dropLength := combinedLen - newLength
+	retainLength := len(bp.flushedSequences) - dropLength
```

Comment:
Thanks, personally I feel like the extra vars and comments have made this MUCH easier to follow.